### PR TITLE
Always execute callback in withScope

### LINKED ISSUE #3030: `withScope()` does not get executed when Sentry is disabled
### Integration

sentry

### Java Version

17

### Version

6.33.0

### Steps to Reproduce

1. Paste the following code into your editor with Sentry in scope
```java
import io.sentry.Sentry;

public class Repro {
  public static void main(String[] args) {
    withoutSentry();
    withSentryDisabled();
    withSentry();
  }

  private static void withoutSentry() {
    System.out.println("without sentry:");

    System.out.println("pre Sentry.withScope");
    Sentry.withScope(s -> {
      System.out.println("in Sentry.withScope"); // <- this will never run
    });
  }

  private static void withSentryDisabled() {
    System.out.println("====\nwith sentry disabled:");

    Sentry.init(o -> o.setEnabled(false));

    System.out.println("pre Sentry.withScope");
    Sentry.withScope(s -> {
      System.out.println("in Sentry.withScope"); // <- this will never run
    });
  }

  private static void withSentry() {
    System.out.println("====\nwith sentry:");

    Sentry.init(o -> o.setDsn("<SOME VALID DSN>"));

    System.out.println("pre Sentry.withScope");
    Sentry.withScope(s -> {
      System.out.println("in Sentry.withScope"); // <- this will
    });
  }
}
```
2. Add a valid DSN into the last function

### Expected Result

In all cases, the `in Sentry.withScope` messages get logged.

### Actual Result

It only logs the message when sentry is properly configured:
image

The docs don't say anything about this behaviour.

### PR DESCRIPTION
**Pull request by lbloder**
## :scroll: Description
Introduced `IScope` and `NoOpScope`.
Replaced all usages of `Scope` types with `IScope`.
Introduced `clone` method as a replacement for the `Scope(Scope scope)` constructor
All `Scope` methods with `default` visibility were changed to `public` and marked as `@ApiStatus.Internal`

## :bulb: Motivation and Context
Fixes #3030

Allow `withScope` to execute the `ScopeCallback` also in the following Scenarios:
- Sentry not initalized
- Sentry disabled through options
- Hub closed

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

**Review by adinauer — APPROVED** [review]
Mostly LGTM. Some changes are probably not intended for this PR. There's some more callbacks but I guess we can ignore them since they're marked `@Internal`.

File: sentry-spring/src/main/java/io/sentry/spring/graphql/SentryGraphqlBeanPostProcessor.java [hunk 1]
```diff
@@ -4,8 +4,10 @@
 import org.springframework.beans.BeansException;
 import org.springframework.beans.factory.config.BeanPostProcessor;
 import org.springframework.graphql.execution.BatchLoaderRegistry;
+import org.springframework.stereotype.Component;
```
[the remark this answers is not in the document]
Comment (lbloder) [reply sentry-spring/src/main/java/io/sentry/spring/graphql/SentryGraphqlBeanPostProcessor.java:7]: yep, my bad, will remove

File: sentry-spring/src/main/java/io/sentry/spring/checkin/SentryCheckInAdviceConfiguration.java [hunk 2]
```diff
@@ -11,18 +11,22 @@
 import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Lazy;
```
[the remark this answers is not in the document]
Comment (lbloder) [reply sentry-spring/src/main/java/io/sentry/spring/checkin/SentryCheckInAdviceConfiguration.java:14]: yep, my bad, will remove

File: sentry/src/main/java/io/sentry/NoOpScope.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    return new ArrayList<>();
+  }
+
+  @Override
+  public void addEventProcessor(@NotNull EventProcessor eventProcessor) {}
+
+  @ApiStatus.Internal
+  @Override
+  public @Nullable Session withSession(Scope.@NotNull IWithSession sessionCallback) {
+    return null;
```
[the remark this answers is not in the document]
Comment (lbloder) [reply sentry/src/main/java/io/sentry/NoOpScope.java:191]: keeping as is for now

File: sentry/src/main/java/io/sentry/NoOpScope.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+  @Override
+  public @NotNull PropagationContext getPropagationContext() {
+    return new PropagationContext();
+  }
+
+  @ApiStatus.Internal
+  @Override
+  public @NotNull PropagationContext withPropagationContext(
+      Scope.@NotNull IWithPropagationContext callback) {
+    return new PropagationContext();
```
[the remark this answers is not in the document]
Comment (lbloder) [reply sentry/src/main/java/io/sentry/NoOpScope.java:236]: keeping as is for now

File: sentry/src/main/java/io/sentry/NoOpScope.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+  @ApiStatus.Internal
+  @Override
+  public @Nullable Session endSession() {
+    return null;
+  }
+
+  @ApiStatus.Internal
+  @Override
+  public void withTransaction(Scope.@NotNull IWithTransaction callback) {}
```
[the remark this answers is not in the document]
Comment (lbloder) [reply sentry/src/main/java/io/sentry/NoOpScope.java:208]: keeping as is for now

**Pull request merged**